lib/semantics/resolve-labels.cc

```diff
@@ -0,0 +1,969 @@
+  void CheckLabelContext(const char *const stmtString, const A &constructName) {
+    const auto I{std::find(
+        constructNames_.crbegin(), constructNames_.crend(), constructName)};
+    if (I == constructNames_.crend()) {
+      Report(eh, currentPosition_,
+          parser::MessageFormattedText{
+              "%s construct-name '%s' is not in scope"_err_en_US, stmtString,
+              constructName.c_str()});
+    }
+  }
+
+  /// \brief Check label range
+  /// Constraint per section 6.2.5, paragraph 2
+  void CheckLabelInRange(parser::Label label) {
+    if ((label < 1) || (label > 99999)) {
+      // this is an error: labels must have a value 1 to 99999, inclusive
+      Report(eh, currentPosition_,
+          parser::MessageFormattedText{
+              "label '%lu' is out of range"_err_en_US, label});
+    }
+  }
+
+  /// \brief Add a labeled statement (label must be distinct)
+  /// Constraint per section 6.2.5., paragraph 2
+  void AddTrgt(parser::Label label, TargetStmtType targetStmtType) {
+    CheckLabelInRange(label);
+    const auto pair{programUnits_.back().targetStmts_.insert(
+        {label, {currentScope_, currentPosition_, targetStmtType}})};
+    if (!pair.second) {
+      // this is an error: labels must be pairwise distinct
+      Report(eh, currentPosition_,
+          parser::MessageFormattedText{
+              "label '%lu' is not distinct"_err_en_US, label});
+    }
+    // Don't enforce a limit to the cardinality of labels
+  }
+
+  /// \brief Reference to a labeled statement from a DO statement
+  void AddDoBase(parser::Label label) {
+    CheckLabelInRange(label);
+    programUnits_.back().doStmtSources_.push_back(
+        {label, currentScope_, currentPosition_});
+  }
+
+  /// \brief Reference to a labeled FORMAT statement
+  void AddFmtBase(parser::Label label) {
+    CheckLabelInRange(label);
+    programUnits_.back().formatStmtSources_.push_back(
+        {label, currentScope_, currentPosition_});
+  }
+
+  /// \brief Reference to a labeled statement as a (possible) branch
+  void AddBase(parser::Label label) {
+    CheckLabelInRange(label);
+    programUnits_.back().otherStmtSources_.push_back(
+        {label, currentScope_, currentPosition_});
+  }
+
+  /// \brief References to labeled statements as (possible) branches
+  void AddBase(const std::list<parser::Label> &labels) {
+    for (const parser::Label &label : labels) {
+      AddBase(label);
+    }
+  }
+
+  std::vector<UnitAnalysis> programUnits_;  ///< results for each program unit
+  ErrorHandler eh;  ///< error handler, collects messages
+  parser::CharBlock currentPosition_{
+      nullptr};  ///< current location in parse tree
+  ScopeProxy currentScope_{0};  ///< current scope in the model
+  std::vector<std::string> constructNames_;
+};
+
+template<typename A, typename B>
+bool InInclusiveScope(const A &scopes, B tail, const B &head) {
+  assert(HasScope(head));
+  assert(HasScope(tail));
+  while (HasScope(tail) && (tail != head)) {
```

Comment:
Redundant parentheses.  Also, the flow is somewhat unclear.  Assuming that `HasScope` is pure, consider this:

```
for (; tail != head; tail = scopes[tail]) {
  if (!HasScope(tail)) {
    return false;
  }
}
return true;
```

Reply:
syntax changes made